`Any` apparently not considered part of `Union[Any, builtins.None]`

When running `mypy --strict` against the following code:

```py
from typing import Any, Optional

class Test(object):

    def __init__(self) -> None:
        self.attr = "foo"  # type: Any

    def foo(self, do_it: bool) -> Optional[Any]:
        if do_it:
            return self.attr
        else:
            return None
```

I see the following warning:

```
testcase.py:11: warning: Returning Any from function with declared return type "Union[Any, builtins.None]"
```

This seems obviously incorrect. At the very least, it's a confusing error message.

Bisecting suggests that this behaviour was introduced in 4141d2039c709f3a8e591fa35a560e54c4e203d2.

--- a/mypy/checker.py
+++ b/mypy/checker.py
@@ -1874,7 +1874,8 @@ class TypeChecker(NodeVisitor[None]):
                 if isinstance(typ, AnyType):
                     # (Unless you asked to be warned in that case, and the
                     # function is not declared to return Any)
-                    if not isinstance(return_type, AnyType) and self.options.warn_return_any:
+                    if (not is_proper_subtype(AnyType(), return_type) and
+                            self.options.warn_return_any):
                         self.warn(messages.RETURN_ANY.format(return_type), s)
                     return
 

--- a/test-data/unit/check-warnings.test
+++ b/test-data/unit/check-warnings.test
@@ -165,3 +165,20 @@ from typing import Any
 def g() -> Any: pass
 def f() -> Any: return g()
 [out]
+
+[case testOKReturnAnyIfProperSubtype]
+# flags: --warn-return-any --strict-optional
+from typing import Any, Optional
+
+class Test(object):
+
+    def __init__(self) -> None:
+        self.attr = "foo"  # type: Any
+
+    def foo(self, do_it: bool) -> Optional[Any]:
+        if do_it:
+            return self.attr  # Should not warn here
+        else:
+            return None
+[builtins fixtures/list.pyi]
+[out]

